File: src/walter/package.py
"""High-level Walter build/create/delete workflows and package entry point."""
from __future__ import annotations

from big_config import ENV
from big_config import workflow as bc_workflow
from big_config.core import Opts, StepFn, workflow
from big_config.step_fns import exit_step_fn, print_error_step_fn
from once import tools as once_tools

from . import tools as walter_tools
from .interop import PARAMS, sync_aliases, to_bc_opts
from .params import opts_fn
from .validation import validate

START = "io.github.bigconfig-ai.walter.package/start"
END = "io.github.bigconfig-ai.walter.package/end"
PIPELINE_START = "io.github.bigconfig-ai.walter.package/start-create-or-delete"
PIPELINE_END = "io.github.bigconfig-ai.walter.package/end-create-or-delete"

step_fns: list[StepFn] = [bc_workflow.print_step_fn, exit_step_fn(END), print_error_step_fn(END)]

TOFU_APPLY = "render tofu:init tofu:apply:-auto-approve"
TOFU_DESTROY = "render tofu:init tofu:destroy:-auto-approve"
ANSIBLE_RUN = "render ansible-playbook:main.yml"

PIPELINE_TOOLS = [
    (once_tools.TOFU, once_tools.tofu),
    (walter_tools.ANSIBLE, walter_tools.ansible),
    (once_tools.ANSIBLE_LOCAL, once_tools.ansible_local),
]

for step, fn in PIPELINE_TOOLS:
    bc_workflow.register_workflow_step(step, fn)

create = bc_workflow.workflow_star(
    {
        "first_step": PIPELINE_START,
        "last_step": PIPELINE_END,
        "pipeline": [
            once_tools.TOFU,
            [TOFU_APPLY, opts_fn],
            walter_tools.ANSIBLE,
            [ANSIBLE_RUN, opts_fn],
            once_tools.ANSIBLE_LOCAL,
            [ANSIBLE_RUN, opts_fn],
        ],
    }
)

build = bc_workflow.workflow_star(
    {
        "first_step": PIPELINE_START,
        "last_step": PIPELINE_END,
        "pipeline": [
            once_tools.TOFU,
            ["render", opts_fn],
            walter_tools.ANSIBLE,
            ["render", opts_fn],
            once_tools.ANSIBLE_LOCAL,
            ["render", opts_fn],
        ],
    }
)

delete_workflow = bc_workflow.workflow_star(
    {
        "first_step": PIPELINE_START,
        "last_step": PIPELINE_END,
        "pipeline": [once_tools.TOFU, [TOFU_DESTROY, opts_fn]],
    }
)

TOOL_OPTS_KEYS = [
    f"{once_tools.TOFU}-opts",
    f"{walter_tools.ANSIBLE}-opts",
    f"{once_tools.ANSIBLE_LOCAL}-opts",
]


def walter(sfns: list[StepFn], opts: Opts) -> Opts:
    """Run a validate / build / create / delete workflow."""
    opts = to_bc_opts(opts)
    with_fns: Opts = {
        bc_workflow.CREATE_FN: create,
        bc_workflow.BUILD_FN: build,
        bc_workflow.DELETE_FN: delete_workflow,
        bc_workflow.VALIDATE_FN: validate,
        **opts,
    }
    merged = bc_workflow.merge_params(TOOL_OPTS_KEYS, opts.get(PARAMS) or {}, with_fns)

    def wire(step: str, resolved_step_fns: list[StepFn]):
        if step == START:
            return (lambda o: bc_workflow.run_steps(resolved_step_fns, o)), END
        return (lambda o: o), None

    wf = workflow({"first_step": START, "wire_fn": wire})
    return sync_aliases(wf(sfns, merged))


def walter_star(args: str | list[str], opts: Opts | None = None) -> Opts:
    """CLI-ready package entry point."""
    parsed = bc_workflow.parse_args(args)
    return walter(step_fns, {**parsed, ENV: "shell", **to_bc_opts(opts or {})})


# Compatibility aliases.
deleteWorkflow = delete_workflow
walterStar = walter_star

File: src/clj/io/github/bigconfig_ai/walter/package.clj
(ns io.github.bigconfig-ai.walter.package
  (:require
   [big-config :as bc]
   [big-config.core :as core]
   [big-config.run :as run]
   [big-config.step-fns :as step-fns]
   [big-config.utils :refer [debug]]
   [big-config.workflow :as workflow]
   [io.github.bigconfig-ai.once.tools :as tools-once]
   [io.github.bigconfig-ai.walter.describe :as describe]
   [io.github.bigconfig-ai.walter.options :as options]
   [io.github.bigconfig-ai.walter.params :as params]
   [io.github.bigconfig-ai.walter.tools :as tools-walter]
   [io.github.bigconfig-ai.walter.validation :as validation]))

(def step-fns [workflow/print-step-fn
               (step-fns/->exit-step-fn ::end)
               (step-fns/->print-error-step-fn ::end)])

(def create
  (workflow/->workflow* {:first-step ::start-create-or-delete
                         :last-step ::end-create-or-delete
                         :pipeline [::tools-once/tofu ["render tofu:init tofu:apply:-auto-approve" params/opts-fn]
                                    ::tools-walter/ansible ["render ansible-playbook:main.yml" params/opts-fn]
                                    ::tools-once/ansible-local ["render ansible-playbook:main.yml" params/opts-fn]]}))

(def build
  (workflow/->workflow* {:first-step ::start-create-or-delete
                         :last-step ::end-create-or-delete
                         :pipeline [::tools-once/tofu ["render" params/opts-fn]
                                    ::tools-walter/ansible ["render" params/opts-fn]
                                    ::tools-once/ansible-local ["render" params/opts-fn]]}))

(def delete
  (workflow/->workflow* {:first-step ::start-create-or-delete
                         :last-step ::end-create-or-delete
                         :pipeline [::tools-once/tofu ["render tofu:init tofu:destroy:-auto-approve" params/opts-fn]]}))

(def ^:private tool-workflows
  {::tools-once/tofu tools-once/tofu
   ::tools-walter/ansible tools-walter/ansible
   ::tools-once/ansible-local tools-once/ansible-local})

(when-let [register-workflow (ns-resolve 'big-config.workflow 'register-workflow)]
  (run! (fn [[step f]]
          (register-workflow step f))
        tool-workflows))

(comment
  (debug tap-values
    (create [] (merge options/walter
                      {::bc/env :repl
                       ::tools-once/tofu-opts (workflow/parse-args "render")
                       ::tools-walter/ansible-opts (workflow/parse-args "render")
                       ::tools-once/ansible-local-opts (workflow/parse-args "render")
                       ::run/shell-opts {:err *err*
                                         :out *out*}})))
  (-> tap-values))

(defn walter
  [step-fns {:keys [::workflow/params] :as opts}]
  (let [opts (->> opts
                  (merge {::workflow/create-fn create
                          ::workflow/build-fn build
                          ::workflow/delete-fn delete
                          ::workflow/validate-fn validation/validate
                          ::workflow/describe-fn describe/describe})
                  (workflow/merge-params [::tools-once/tofu-opts
                                          ::tools-walter/ansible-opts
                                          ::tools-once/ansible-local-opts]
                                         params))
        wf (core/->workflow {:first-step ::start
                             :wire-fn (fn [step step-fns]
                                        (case step
                                          ::start [(partial workflow/run-steps step-fns) ::end]
                                          ::end [identity]))})]
    (wf step-fns opts)))

(defn walter*
  [args & [opts]]
  (let [opts (merge (workflow/parse-args args)
                    {::bc/env :shell}
                    opts)]
    (walter step-fns opts)))

(comment
  (debug tap-values
    (walter* "build" (merge options/walter
                            {::bc/env :repl
                             ::run/shell-opts {:err *err*
                                               :out *out*}})))
  (-> tap-values))
